File: frontend/lib/api/activities.ts
import { apiFetch } from "@/lib/api/client";
import type { Activity, ActivityFilters, ActivityListResponse } from "@/types/api";

export type ActivityInput = {
  entity_type: string;
  entity_id: string;
  activity_type: string;
  description: string;
  title?: string | null;
  action?: string | null;
  metadata?: Record<string, unknown> | null;
  scheduled_at?: string | null;
};

export const ACTIVITY_CATEGORIES = [
  { value: "", label: "All" },
  { value: "deals", label: "Deals" },
  { value: "companies", label: "Companies" },
  { value: "contacts", label: "Contacts" },
  { value: "tasks", label: "Tasks" },
  { value: "notes", label: "Notes" },
  { value: "authentication", label: "Authentication" },
] as const;

export const ACTIVITY_TYPES = [
  "call", "meeting", "email", "note",
  "company_created", "company_updated", "company_deleted",
  "contact_created", "contact_updated", "contact_deleted",
  "lead_created", "lead_updated", "lead_assigned", "lead_converted", "lead_deleted",
  "deal_created", "deal_updated", "deal_stage_changed", "deal_moved", "deal_won", "deal_lost", "deal_deleted",
  "task_created", "task_updated", "task_completed", "task_reopened", "task_deleted",
  "note_added", "note_edited",
  "user_login", "user_invited", "password_reset",
] as const;

export const ENTITY_TYPES = ["lead", "contact", "deal", "company", "task", "user", "tenant"] as const;

export const ACTIVITY_TYPE_LABELS: Record<string, string> = {
  call: "Call", meeting: "Meeting", email: "Email", note: "Note",
  company_created: "Company Created", company_updated: "Company Updated", company_deleted: "Company Deleted",
  contact_created: "Contact Created", contact_updated: "Contact Updated", contact_deleted: "Contact Deleted",
  lead_created: "Lead Created", lead_updated: "Lead Updated", lead_assigned: "Lead Assigned",
  lead_converted: "Lead Converted", lead_deleted: "Lead Deleted",
  deal_created: "Deal Created", deal_updated: "Deal Updated", deal_stage_changed: "Stage Changed",
  deal_moved: "Deal Moved", deal_won: "Deal Won", deal_lost: "Deal Lost", deal_deleted: "Deal Deleted",
  task_created: "Task Created", task_updated: "Task Updated", task_completed: "Task Completed",
  task_reopened: "Task Reopened", task_deleted: "Task Deleted",
  note_added: "Note Added", note_edited: "Note Edited",
  user_login: "Login", user_invited: "User Invited", password_reset: "Password Reset",
};

function buildQuery(filters: ActivityFilters): string {
  const params = new URLSearchParams();
  if (filters.q) params.set("q", filters.q);
  if (filters.entity_type) params.set("entity_type", filters.entity_type);
  if (filters.entity_id) params.set("entity_id", filters.entity_id);
  if (filters.activity_type) params.set("activity_type", filters.activity_type);
  if (filters.action) params.set("action", filters.action);
  if (filters.actor_id) params.set("actor_id", filters.actor_id);
  if (filters.date_from) params.set("date_from", filters.date_from);
  if (filters.date_to) params.set("date_to", filters.date_to);
  if (filters.category) params.set("category", filters.category);
  if (filters.cursor) params.set("cursor", filters.cursor);
  if (filters.sort) params.set("sort", filters.sort);
  if (filters.page) params.set("page", String(filters.page));
  if (filters.page_size) params.set("page_size", String(filters.page_size));
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function listActivities(
  slug: string,
  filters: ActivityFilters = {},
): Promise<ActivityListResponse> {
  return apiFetch<ActivityListResponse>(`/tenants/${slug}/activities${buildQuery(filters)}`);
}

export async function listLeadActivities(
  slug: string,
  leadId: string,
  filters: Omit<ActivityFilters, "entity_type" | "entity_id"> = {},
): Promise<ActivityListResponse> {
  return apiFetch<ActivityListResponse>(`/tenants/${slug}/activities/lead/${leadId}${buildQuery(filters)}`);
}

export async function listContactActivities(
  slug: string,
  contactId: string,
  filters: Omit<ActivityFilters, "entity_type" | "entity_id"> = {},
): Promise<ActivityListResponse> {
  return apiFetch<ActivityListResponse>(`/tenants/${slug}/activities/contact/${contactId}${buildQuery(filters)}`);
}

export async function listDealActivities(
  slug: string,
  dealId: string,
  filters: Omit<ActivityFilters, "entity_type" | "entity_id"> = {},
): Promise<ActivityListResponse> {
  return apiFetch<ActivityListResponse>(`/tenants/${slug}/activities/deal/${dealId}${buildQuery(filters)}`);
}

export async function listCompanyActivities(
  slug: string,
  companyId: string,
  filters: Omit<ActivityFilters, "entity_type" | "entity_id"> = {},
): Promise<ActivityListResponse> {
  return apiFetch<ActivityListResponse>(`/tenants/${slug}/activities/company/${companyId}${buildQuery(filters)}`);
}

export async function getActivity(slug: string, activityId: string): Promise<Activity> {
  return apiFetch<Activity>(`/tenants/${slug}/activities/${activityId}`);
}

export async function createActivity(slug: string, data: ActivityInput): Promise<Activity> {
  return apiFetch<Activity>(`/tenants/${slug}/activities`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateActivity(
  slug: string,
  activityId: string,
  data: Partial<ActivityInput>,
): Promise<Activity> {
  return apiFetch<Activity>(`/tenants/${slug}/activities/${activityId}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

export async function deleteActivity(slug: string, activityId: string): Promise<void> {
  await apiFetch<void>(`/tenants/${slug}/activities/${activityId}`, { method: "DELETE" });
}

export async function bulkDeleteActivities(slug: string, ids: string[]): Promise<{ affected: number }> {
  return apiFetch<{ affected: number }>(`/tenants/${slug}/activities/bulk-delete`, {
    method: "POST",
    body: JSON.stringify({ ids }),
  });
}

export async function bulkArchiveActivities(slug: string, ids: string[]): Promise<{ affected: number }> {
  return apiFetch<{ affected: number }>(`/tenants/${slug}/activities/bulk-archive`, {
    method: "POST",
    body: JSON.stringify({ ids }),
  });
}

export function formatRelativeTime(value: string): string {
  const date = new Date(value);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffMins = Math.floor(diffMs / 60000);
  const diffHours = Math.floor(diffMs / 3600000);
  const diffDays = Math.floor(diffMs / 86400000);

  if (diffMins < 1) return "Just now";
  if (diffMins < 60) return `${diffMins}m ago`;
  if (diffHours < 24) return `${diffHours}h ago`;
  if (diffDays === 1) return "Yesterday";
  if (diffDays < 7) return `${diffDays}d ago`;

  return date.toLocaleDateString("en-US", { month: "short", day: "numeric" });
}

export function getActivityTitle(activity: Activity): string {
  return activity.title || ACTIVITY_TYPE_LABELS[activity.action] || activity.action.replace(/_/g, " ");
}

// Legacy emoji map for embedded timelines
export const ACTIVITY_TYPE_ICONS: Record<string, string> = {
  call: "📞", meeting: "📅", email: "📧", note: "📝",
  deal_created: "✨", deal_moved: "↔️", deal_updated: "✏️",
  deal_deleted: "🗑️", deal_won: "🏆", deal_lost: "❌",
};

File: backend/app/api/v1/activities.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import TenantContext, require_permission
from app.db.session import get_db
from app.schemas.activity import (
    ActivityBulkIds,
    ActivityBulkResult,
    ActivityCreate,
    ActivityEntityRef,
    ActivityListResponse,
    ActivityResponse,
    ActivityUpdate,
)
from app.services.activity_service import ActivityService, paginate

router = APIRouter(prefix="/tenants/{slug}/activities", tags=["activities"])


def _actor_dict(user) -> dict | None:
    if not user:
        return None
    return {"id": user.id, "full_name": user.full_name, "email": user.email}


def _to_response(activity, entity_ref=None) -> ActivityResponse:
    actor = _actor_dict(activity.created_by)
    entity = None
    if entity_ref:
        entity = ActivityEntityRef(
            entity_type=entity_ref.entity_type,
            entity_id=entity_ref.entity_id,
            display_name=entity_ref.display_name,
            href_path=entity_ref.href_path,
        )
    return ActivityResponse(
        id=activity.id,
        tenant_id=activity.tenant_id,
        entity_type=activity.entity_type,
        entity_id=activity.entity_id,
        activity_type=activity.activity_type,
        action=activity.action,
        title=activity.title,
        description=activity.description,
        icon=activity.icon,
        color=activity.color,
        metadata=activity.activity_metadata,
        actor_id=activity.created_by_id,
        actor=actor,
        created_by_id=activity.created_by_id,
        created_by=actor,
        entity=entity,
        created_at=activity.created_at,
        scheduled_at=activity.scheduled_at,
        archived_at=activity.archived_at,
    )


def _list_response(
    service: ActivityService,
    ctx: TenantContext,
    activities: list,
    total: int,
    *,
    page: int | None = None,
    page_size: int | None = None,
    next_cursor: str | None = None,
    has_more: bool = False,
) -> ActivityListResponse:
    refs = service.resolve_entities(ctx.tenant.id, ctx.tenant.slug, activities)
    items = [
        _to_response(a, refs.get((a.entity_type, a.entity_id))) for a in activities
    ]
    meta = paginate(total, page, page_size) if page and page_size else {}
    return ActivityListResponse(
        items=items,
        total=total,
        page=meta.get("page"),
        page_size=meta.get("page_size"),
        pages=meta.get("pages"),
        next_cursor=next_cursor,
        has_more=has_more,
    )


@router.get("", response_model=ActivityListResponse)
def list_activities(
    q: str | None = Query(default=None, max_length=200),
    entity_type: str | None = Query(default=None, max_length=30),
    entity_id: UUID | None = Query(default=None),
    activity_type: str | None = Query(default=None, max_length=50),
    action: str | None = Query(default=None, max_length=50),
    actor_id: UUID | None = Query(default=None),
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    category: str | None = Query(default=None, max_length=30),
    page: int | None = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    cursor: str | None = Query(default=None),
    sort: str = Query(default="desc", pattern="^(asc|desc)$"),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityListResponse:
    service = ActivityService(db)
    if cursor:
        activities, total, next_cursor, has_more = service.list_activities(
            ctx.tenant.id,
            q=q,
            entity_type=entity_type,
            entity_id=entity_id,
            activity_type=activity_type,
            action=action,
            actor_id=actor_id,
            date_from=date_from,
            date_to=date_to,
            category=category,
            page_size=page_size,
            cursor=cursor,
            sort=sort,
        )
        return _list_response(
            service, ctx, activities, total,
            page_size=page_size, next_cursor=next_cursor, has_more=has_more,
        )

    activities, total, _, _ = service.list_activities(
        ctx.tenant.id,
        q=q,
        entity_type=entity_type,
        entity_id=entity_id,
        activity_type=activity_type,
        action=action,
        actor_id=actor_id,
        date_from=date_from,
        date_to=date_to,
        category=category,
        page=page or 1,
        page_size=page_size,
        sort=sort,
    )
    return _list_response(
        service, ctx, activities, total, page=page or 1, page_size=page_size,
    )


@router.post("", response_model=ActivityResponse, status_code=201)
def create_activity(
    payload: ActivityCreate,
    ctx: TenantContext = Depends(require_permission("activity:write")),
    db: Session = Depends(get_db),
) -> ActivityResponse:
    service = ActivityService(db)
    activity = service.create_activity(
        ctx.tenant.id, payload, ctx.membership.user_id
    )
    refs = service.resolve_entities(ctx.tenant.id, ctx.tenant.slug, [activity])
    return _to_response(activity, refs.get((activity.entity_type, activity.entity_id)))


@router.post("/bulk-delete", response_model=ActivityBulkResult)
def bulk_delete_activities(
    payload: ActivityBulkIds,
    ctx: TenantContext = Depends(require_permission("activity:delete")),
    db: Session = Depends(get_db),
) -> ActivityBulkResult:
    count = ActivityService(db).bulk_delete(ctx.tenant.id, payload.ids)
    return ActivityBulkResult(affected=count)


@router.post("/bulk-archive", response_model=ActivityBulkResult)
def bulk_archive_activities(
    payload: ActivityBulkIds,
    ctx: TenantContext = Depends(require_permission("activity:write")),
    db: Session = Depends(get_db),
) -> ActivityBulkResult:
    count = ActivityService(db).bulk_archive(ctx.tenant.id, payload.ids)
    return ActivityBulkResult(affected=count)


@router.get("/lead/{lead_id}", response_model=ActivityListResponse)
def list_lead_activities(
    lead_id: UUID,
    q: str | None = Query(default=None, max_length=200),
    activity_type: str | None = Query(default=None, max_length=50),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityListResponse:
    service = ActivityService(db)
    activities, total, _, _ = service.list_activities(
        ctx.tenant.id,
        q=q,
        entity_type="lead",
        entity_id=lead_id,
        activity_type=activity_type,
        page=page,
        page_size=page_size,
    )
    return _list_response(service, ctx, activities, total, page=page, page_size=page_size)


@router.get("/contact/{contact_id}", response_model=ActivityListResponse)
def list_contact_activities(
    contact_id: UUID,
    q: str | None = Query(default=None, max_length=200),
    activity_type: str | None = Query(default=None, max_length=50),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityListResponse:
    service = ActivityService(db)
    activities, total, _, _ = service.list_activities(
        ctx.tenant.id,
        q=q,
        entity_type="contact",
        entity_id=contact_id,
        activity_type=activity_type,
        page=page,
        page_size=page_size,
    )
    return _list_response(service, ctx, activities, total, page=page, page_size=page_size)


@router.get("/deal/{deal_id}", response_model=ActivityListResponse)
def list_deal_activities(
    deal_id: UUID,
    q: str | None = Query(default=None, max_length=200),
    activity_type: str | None = Query(default=None, max_length=50),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityListResponse:
    service = ActivityService(db)
    activities, total, _, _ = service.list_activities(
        ctx.tenant.id,
        q=q,
        entity_type="deal",
        entity_id=deal_id,
        activity_type=activity_type,
        page=page,
        page_size=page_size,
    )
    return _list_response(service, ctx, activities, total, page=page, page_size=page_size)


@router.get("/company/{company_id}", response_model=ActivityListResponse)
def list_company_activities(
    company_id: UUID,
    q: str | None = Query(default=None, max_length=200),
    activity_type: str | None = Query(default=None, max_length=50),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityListResponse:
    service = ActivityService(db)
    activities, total, _, _ = service.list_activities(
        ctx.tenant.id,
        q=q,
        entity_type="company",
        entity_id=company_id,
        activity_type=activity_type,
        page=page,
        page_size=page_size,
    )
    return _list_response(service, ctx, activities, total, page=page, page_size=page_size)


@router.get("/{activity_id}", response_model=ActivityResponse)
def get_activity(
    activity_id: UUID,
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> ActivityResponse:
    service = ActivityService(db)
    activity = service.get_activity(ctx.tenant.id, activity_id)
    refs = service.resolve_entities(ctx.tenant.id, ctx.tenant.slug, [activity])
    return _to_response(activity, refs.get((activity.entity_type, activity.entity_id)))


@router.patch("/{activity_id}", response_model=ActivityResponse)
def update_activity(
    activity_id: UUID,
    payload: ActivityUpdate,
    ctx: TenantContext = Depends(require_permission("activity:write")),
    db: Session = Depends(get_db),
) -> ActivityResponse:
    service = ActivityService(db)
    activity = service.update_activity(ctx.tenant.id, activity_id, payload)
    refs = service.resolve_entities(ctx.tenant.id, ctx.tenant.slug, [activity])
    return _to_response(activity, refs.get((activity.entity_type, activity.entity_id)))


@router.delete("/{activity_id}", status_code=204)
def delete_activity(
    activity_id: UUID,
    ctx: TenantContext = Depends(require_permission("activity:delete")),
    db: Session = Depends(get_db),
) -> None:
    ActivityService(db).delete_activity(ctx.tenant.id, activity_id)
